sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs

```diff
@@ -167,24 +180,34 @@ public Task StartAsync(CancellationToken cancellationToken)
             return Task.FromResult(0);
         }
 
-        public async Task StopAsync(CancellationToken cancellationToken)
-        {
-            if (!_drainModeManager?.IsDrainModeEnabled ?? true)
-            {
-                // Cancel the execution token when drain mode is not enabled or drain mode manager is not set.
-                _executionCancellationTokenSource.Cancel();
-            }
-
-            using (cancellationToken.Register(() => _shutdownCancellationTokenSource.Cancel()))
-            {
-                ThrowIfDisposed();
-                _timer.Cancel();
-                await Task.WhenAll(_processing).ConfigureAwait(false);
-                await _timer.StopAsync(cancellationToken).ConfigureAwait(false);
-                _logger.LogDebug($"Storage queue listener stopped ({_details})");
-            }
+       public async Task StopAsync(CancellationToken cancellationToken)
+       {
+           if (_drainModeManager == null || !_drainModeManager.IsDrainModeEnabled)
+           {
+               _logger.LogDebug($"Drain mode is not enabled, storage queue listener will stop immediately ({_details})");
+               // Non-drain mode: cancel execution and link host cancellation to shutdown
+               _executionCancellationTokenSource.Cancel();
+               using (cancellationToken.Register(() => _shutdownCancellationTokenSource.Cancel()))
+               {
+                   ThrowIfDisposed();
+                   _timer.Cancel();
+                   await Task.WhenAll(_processing).ConfigureAwait(false);
+                   await _timer.StopAsync(cancellationToken).ConfigureAwait(false);
+                   _logger.LogDebug($"Drain mode is not enabled, storage queue listener stopped ({_details})");
```

Comment:
This is introducing a lot of code duplication between these two branches. Instead, just isolate the lines that differ in your if/else block, and have the common code outside.

To simplify, probably you can just remove the "using (cancellationToken.Register" - I don't think the shutdown tasks need to be within a using scope

Reply:
@mathewc How about this refactor? I can also remove the using scope entirely, but this is maybe a good compromise while preserving the existing behavior?